source/docs/h2_metadata.md

```diff
@@ -0,0 +1,106 @@
+### Overview
+
+Envoy provides a way for users to communicate extra information associating with a stream that is
+not carried in the standard HTTP(s) and HTTP/2 headers and payloads. The
+information is represented by a string key-value pair. For example, users can
+pass along RTT information associated with a stream using a key of "rtt info", and a value of
+"100ms". In Envoy, we call this type of information metadata.
+A stream can be associated with multiple metadata, and the multiple metadata
+are represented by a map.
+
+Note: the metadata implementation is still in progress, and the doc is in draft
+version.
+
+### Limitation
+
+For ease of implementation and compatibility purposes, metadata will only be
+supported in HTTP/2. Metadata sent in any other protocol should result in protocol
+errors or be ignored.
+
+### Envoy metadata handling
+
+Envoy provides the functionality to proxy, process and add metadata.
+
+## Proxying metadata
+
+(To be implemented)
+
+If not specified, all the metadata received by Envoy is proxied to the next hop
+unmodified. Note that, we do not guarantee the same frame order will be preserved from
+hop by hop. That is, metadata from upstream at the beginning of a stream can be
+received by the downstream at the end of the stream.
```

Comment:
Worth commenting somewhere that empty maps will not be serialized out, i.e. if you clear the data it will functionally result in metadata being removed for the next hop?